src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,270 @@
+                        try
+                        {
+                            socket.Blocking = false;
+                            socket.Send(Array.Empty<byte>(), 0, SocketFlags.None);
+                        }
+                        catch (Exception)
+                        {
+                        }
+                        finally
+                        {
+                            socket.Blocking = blocking;
+                        }
+                        return socket.Connected;
+                    }
+                    else if (_stream is PipeStream pipeStream)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            // PeekNamedPipe will return false if the pipe is disconnected/broken.
+                            if (!NativeMethods.PeekNamedPipe(pipeStream.SafePipeHandle, null, 0, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero))
+                            {
+                                int err = Marshal.GetLastWin32Error();
+                                if (109 == err) // ERROR_BROKEN_PIPE
+                                {
+                                    return false;
+                                }
+                                throw new InvalidOperationException($"PeekNamedPipe failed: {err}");
+                            }
+                            return true;
+                        }
+                    }
+
+                    throw new InvalidOperationException($"Stream type '{_stream.GetType().FullName}' was not handled.");
+                }
+                finally
+                {
+                    // Set the event again to allow other callers to get the stream.
+                    _streamReady.Set();
+                }
+            }
+        }
+
+        /// <inheritdoc cref="IIpcEndpoint.Connect"/>
+        /// <remarks>
+        /// This will block until the diagnostic stream is provided. This block can happen if
+        /// the stream is acquired previously and the runtime instance has not yet reconnected
+        /// to the reversed diagnostics server.
+        /// </remarks>
+        public Stream Connect()
+        {
+            _streamReady.WaitOne();
+
+            return ExchangeStream(stream: null);
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                ExchangeStream(stream: null)?.Dispose();
+
+                _streamReady.Dispose();
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Updates the endpoint with a new diagnostics stream.
+        /// </summary>
+        internal void SetStream(Stream stream)
+        {
+            ExchangeStream(stream)?.Dispose();
+
+            _streamReady.Set();
+        }
+
+        private Stream ExchangeStream(Stream stream)
+        {
+            lock (_streamSync)
```

Comment:
Does this need locking and Interlocked? Usually it's one or the other.

Reply:
We need locking at least since the CheckTransport method does some non-atomic manipulation and checking on the existing stream. I can remove the Interlocked.Exchange and replace it with it's non-atomic equivalent (since it's already inside the lock).